File: scripts/op-relay.py
#!/usr/bin/env python3
"""1Password CLI secret relay server.

Listens on a local Unix socket and resolves a narrow set of `op` requests on
the client Mac, where the 1Password desktop app can show the biometric prompt.
The socket is intended to be reverse-forwarded over SSH to a trusted host.

Protocol (newline-delimited JSON, one request per connection):

  Request:  {"args": ["read", "op://vault/item/field"]}
  Response: {"exit_code": 0, "stdout": "secret-value", "stderr": ""}
"""

import json
import os
import signal
import socket
import subprocess
import sys
import threading

SOCK_PATH = os.environ.get("OP_RELAY_SOCK", os.path.expanduser("~/.op-relay.sock"))
TCP_PORT = int(os.environ.get("OP_RELAY_PORT", "12321"))
PID_PATH = SOCK_PATH + ".pid"
OP_BIN = os.environ.get("OP_RELAY_BIN", "op")
TIMEOUT = int(os.environ.get("OP_RELAY_TIMEOUT", "60"))
ISOLATE_COMMANDS = os.environ.get("OP_RELAY_ISOLATE_COMMANDS", "1") not in {
    "0",
    "false",
    "False",
}
SIGNOUT_AFTER_READ = os.environ.get("OP_RELAY_SIGNOUT_AFTER_READ", "1") not in {
    "0",
    "false",
    "False",
}
SIGNOUT_ARGS = os.environ.get("OP_RELAY_SIGNOUT_ARGS", "signout --all").split()

ALLOWED_READ_FLAGS = {
    "--account",
    "--cache",
}

ALLOWED_DIRECT_COMMANDS = {
    "read",
    "whoami",
}


def kill_previous() -> None:
    try:
        with open(PID_PATH) as f:
            pid = int(f.read().strip())
        os.kill(pid, signal.SIGTERM)
    except (FileNotFoundError, ValueError, ProcessLookupError, PermissionError):
        pass


def error(message: str) -> dict[str, object]:
    return {"exit_code": 1, "stdout": "", "stderr": f"op-relay: {message}\n"}


def validate_args(args: object) -> list[str]:
    if not args or not isinstance(args, list):
        raise ValueError("missing or invalid 'args'")
    if not all(isinstance(arg, str) for arg in args):
        raise ValueError("'args' must be a list of strings")
    if args[0] not in ALLOWED_DIRECT_COMMANDS:
        raise ValueError(f"unsupported op command: {args[0]}")
    if args[0] == "read":
        validate_read_args(args[1:])
    elif len(args) > 1:
        raise ValueError(f"unsupported arguments for op {args[0]}")
    return args


def validate_read_args(args: list[str]) -> None:
    positional = []
    i = 0
    while i < len(args):
        arg = args[i]
        if arg == "--":
            positional.extend(args[i + 1 :])
            break
        if arg.startswith("--account="):
            i += 1
            continue
        if arg in ALLOWED_READ_FLAGS:
            if arg == "--account":
                i += 2
            else:
                i += 1
            continue
        if arg.startswith("-"):
            raise ValueError(f"unsupported op read flag: {arg}")
        positional.append(arg)
        i += 1

    if len(positional) != 1 or not positional[0].startswith("op://"):
        raise ValueError("op read relay requires exactly one op:// secret reference")


def run_op(args: list[str]) -> dict[str, object]:
    env = os.environ.copy()
    env.pop("OP_SESSION", None)
    result = subprocess.run(
        [OP_BIN] + args,
        capture_output=True,
        text=True,
        timeout=TIMEOUT,
        env=env,
        start_new_session=ISOLATE_COMMANDS,
    )
    stderr = result.stderr
    if SIGNOUT_AFTER_READ and args[0] == "read":
        signout = subprocess.run(
            [OP_BIN] + SIGNOUT_ARGS,
            capture_output=True,
            text=True,
            timeout=TIMEOUT,
        )
        if signout.returncode != 0:
            stderr += signout.stderr or "op-relay: op signout failed\n"
    return {
        "exit_code": result.returncode,
        "stdout": result.stdout,
        "stderr": stderr,
    }


def handle_client(conn: socket.socket) -> None:
    try:
        data = b""
        while b"\n" not in data:
            chunk = conn.recv(4096)
            if not chunk:
                break
            data += chunk

        if not data.strip():
            return

        req = json.loads(data.decode())
        resp = run_op(validate_args(req.get("args")))

        conn.sendall((json.dumps(resp) + "\n").encode())
    except ValueError as e:
        conn.sendall((json.dumps(error(str(e))) + "\n").encode())
    except json.JSONDecodeError:
        err = error("invalid JSON")
        conn.sendall((json.dumps(err) + "\n").encode())
    except subprocess.TimeoutExpired:
        err = error(f"op command timed out ({TIMEOUT}s)")
        conn.sendall((json.dumps(err) + "\n").encode())
    except Exception as e:
        err = {"exit_code": 1, "stdout": "", "stderr": str(e)}
        try:
            conn.sendall((json.dumps(err) + "\n").encode())
        except Exception:
            pass
    finally:
        conn.close()


def accept_loop(srv: socket.socket) -> None:
    while True:
        try:
            conn, _ = srv.accept()
        except OSError:
            break
        threading.Thread(target=handle_client, args=(conn,), daemon=True).start()


def main() -> None:
    kill_previous()

    if os.path.exists(SOCK_PATH):
        os.unlink(SOCK_PATH)

    unix_srv = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    unix_srv.bind(SOCK_PATH)
    os.chmod(SOCK_PATH, 0o600)
    unix_srv.listen(4)

    tcp_srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    tcp_srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    tcp_srv.bind(("127.0.0.1", TCP_PORT))
    tcp_srv.listen(4)

    with open(PID_PATH, "w") as f:
        f.write(str(os.getpid()))

    def shutdown(signum, frame):
        unix_srv.close()
        tcp_srv.close()
        for p in (SOCK_PATH, PID_PATH):
            try:
                os.unlink(p)
            except FileNotFoundError:
                pass
        sys.exit(0)

    signal.signal(signal.SIGINT, shutdown)
    signal.signal(signal.SIGTERM, shutdown)

    threading.Thread(target=accept_loop, args=(tcp_srv,), daemon=True).start()
    accept_loop(unix_srv)


if __name__ == "__main__":
    main()

File: scripts/op-relay.ts
#!/usr/bin/env bun
import { spawn, spawnSync } from "node:child_process";
import {
  existsSync,
  readFileSync,
  rmSync,
  writeFileSync,
  chmodSync,
} from "node:fs";
import net from "node:net";
import os from "node:os";

const sockPath = process.env.OP_RELAY_SOCK ?? `${os.homedir()}/.op-relay.sock`;
const tcpPort = Number.parseInt(process.env.OP_RELAY_PORT ?? "12321", 10);
const pidPath = `${sockPath}.pid`;
const opBin = process.env.OP_RELAY_BIN ?? "op";
const timeout =
  Number.parseInt(process.env.OP_RELAY_TIMEOUT ?? "60", 10) * 1000;
const isolateCommands = !["0", "false", "False"].includes(
  process.env.OP_RELAY_ISOLATE_COMMANDS ?? "1",
);
const signoutAfterRead = !["0", "false", "False"].includes(
  process.env.OP_RELAY_SIGNOUT_AFTER_READ ?? "1",
);
const signoutArgs = (process.env.OP_RELAY_SIGNOUT_ARGS ?? "signout --all")
  .split(/\s+/)
  .filter(Boolean);

const allowedReadFlags = new Set(["--account", "--cache"]);
const allowedDirectCommands = new Set(["read", "whoami"]);

type RelayResponse = { exit_code: number; stdout: string; stderr: string };

function killPrevious() {
  try {
    const pid = Number.parseInt(readFileSync(pidPath, "utf8").trim(), 10);
    if (Number.isFinite(pid)) process.kill(pid, "SIGTERM");
  } catch {}
}

function error(message: string): RelayResponse {
  return { exit_code: 1, stdout: "", stderr: `op-relay: ${message}\n` };
}

function validateArgs(args: unknown): string[] {
  if (!Array.isArray(args) || args.length === 0)
    throw new Error("missing or invalid 'args'");
  if (!args.every((arg) => typeof arg === "string"))
    throw new Error("'args' must be a list of strings");
  if (!allowedDirectCommands.has(args[0]))
    throw new Error(`unsupported op command: ${args[0]}`);
  if (args[0] === "read") validateReadArgs(args.slice(1));
  else if (args.length > 1)
    throw new Error(`unsupported arguments for op ${args[0]}`);
  return args;
}

function validateReadArgs(args: string[]) {
  const positional: string[] = [];
  for (let i = 0; i < args.length; ) {
    const arg = args[i];
    if (arg === "--") {
      positional.push(...args.slice(i + 1));
      break;
    }
    if (arg.startsWith("--account=")) {
      i += 1;
      continue;
    }
    if (allowedReadFlags.has(arg)) {
      i += arg === "--account" ? 2 : 1;
      continue;
    }
    if (arg.startsWith("-"))
      throw new Error(`unsupported op read flag: ${arg}`);
    positional.push(arg);
    i += 1;
  }
  if (positional.length !== 1 || !positional[0].startsWith("op://")) {
    throw new Error(
      "op read relay requires exactly one op:// secret reference",
    );
  }
}

async function runCommand(
  args: string[],
  env = process.env,
): Promise<RelayResponse> {
  return await new Promise((resolve) => {
    let stdout = "";
    let stderr = "";
    const child = spawn(opBin, args, {
      env,
      detached: isolateCommands,
      stdio: ["ignore", "pipe", "pipe"],
    });
    const timer = setTimeout(() => child.kill("SIGTERM"), timeout);
    child.stdout.setEncoding("utf8").on("data", (chunk) => (stdout += chunk));
    child.stderr.setEncoding("utf8").on("data", (chunk) => (stderr += chunk));
    child.on("error", (err) => {
      clearTimeout(timer);
      resolve({ exit_code: 1, stdout, stderr: stderr || `${err.message}\n` });
    });
    child.on("close", (code, signal) => {
      clearTimeout(timer);
      if (signal) resolve(error(`op command timed out (${timeout / 1000}s)`));
      else resolve({ exit_code: code ?? 1, stdout, stderr });
    });
  });
}

async function runOp(args: string[]): Promise<RelayResponse> {
  const env = { ...process.env };
  delete env.OP_SESSION;
  const result = await runCommand(args, env);
  if (signoutAfterRead && args[0] === "read") {
    const signout = spawnSync(opBin, signoutArgs, {
      encoding: "utf8",
      timeout,
    });
    if (signout.status !== 0)
      result.stderr += signout.stderr || "op-relay: op signout failed\n";
  }
  return result;
}

function handleClient(conn: net.Socket) {
  let data = "";
  conn.setEncoding("utf8");
  conn.on("data", async (chunk) => {
    data += chunk;
    if (!data.includes("\n")) return;
    conn.pause();
    try {
      if (data.trim()) {
        const req = JSON.parse(data);
        const resp = await runOp(validateArgs(req.args));
        conn.end(`${JSON.stringify(resp)}\n`);
      } else {
        conn.end();
      }
    } catch (err) {
      conn.end(
        `${JSON.stringify(error(err instanceof SyntaxError ? "invalid JSON" : String((err as Error).message ?? err)))}\n`,
      );
    }
  });
  conn.on("error", () => {});
}

function cleanup(unixServer: net.Server, tcpServer: net.Server) {
  unixServer.close();
  tcpServer.close();
  for (const path of [sockPath, pidPath]) {
    try {
      rmSync(path);
    } catch {}
  }
}

killPrevious();
try {
  if (existsSync(sockPath)) rmSync(sockPath);
} catch {}

const unixServer = net.createServer(handleClient);
const tcpServer = net.createServer(handleClient);

unixServer.listen(sockPath, () => chmodSync(sockPath, 0o600));
tcpServer.listen(tcpPort, "127.0.0.1");
writeFileSync(pidPath, `${process.pid}`);

for (const sig of ["SIGINT", "SIGTERM"] as const) {
  process.on(sig, () => {
    cleanup(unixServer, tcpServer);
    process.exit(0);
  });
}
